Low level interop with Vulkan backend

zarik5 opened the pull request:
Related issues: #3698 #3761 https://github.com/blaind/xrbevy/issues/1

This PR exposes some Vulkan-backend-specific details needed for interop with other low level APIs. In particular this PR aims to implement a minimum viable API to allow OpenXR integration as a separate crate.

This PR is nowhere complete. I opened it to get some feedback; this is my first PR on a project the size of gfx-hal.

This is the first of multiple PRs needed for OpenXR support in wgpu.

PR checklist:
- [ ] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [ ] tested examples with the following backends: Vulkan

EDIT: To make things clear, this is another take on #3761. Me and @blaind are working together on this route.

zarik5 commented:
I see some fields of the backend structs `Instance`, `PhysicalDevice`, `Device` etc. are `pub`. Shouldn't they be `pub(crate)`? If some fields need to be made public (for example as part of this PR) I think getter methods should be preferred.

zarik5 commented:
For now I remove `::raw()` methods. They are not needed for the OpenXR integration and it is not clear what should be exposed for potential integration with other APIs (which might require Vulkan objects created by gfx-hal). This can be deferred to a future PR.

kvark commented on src/backend/vulkan/src/lib.rs:
```diff
@@ -353,6 +354,205 @@ unsafe extern "system" fn debug_report_callback(
+    /// `raw_instance` must be manually destroyed *after* gfx-hal Instance has been dropped.
+    pub unsafe fn from_raw(
+        #[cfg(not(feature = "use-rtld-next"))] entry: Entry,
+        #[cfg(feature = "use-rtld-next")] entry: EntryCustom<()>,
```
we should probably use the fact that `type Entry = EntryCustom<Arc<Library>>`. So the only thing different between `use-rtld-next` and not is the generic parameter (which we can typedef somewhere).

zarik5 replied:
Great idea! Using cfg-gated parameters was also confusing rust-analyzer.

zarik5 replied:
I preferred using a typedef on `Entry`, typedef-ing only the type argument felt like splitting the type in half.

kvark reviewed (approved):
I really like the way this is written. Top-notch work here!
Assuming this API works for you, let's make this mergeable.

zarik5 commented:
@kvark Thank you! The work here is not finished, I'm developing this API while working on a minimum example for integration with OpenXR. I have to work around the missing support of multiview in gfx-hal, which should be handled in another PR.

zarik5 commented on src/backend/vulkan/src/device.rs:
```diff
@@ -2044,6 +2019,45 @@ impl super::Device {
         Ok((swapchain, images))
     }
+
+    pub unsafe fn image_view_from_raw(
+        &self,
+        raw_image: vk::Image,
+        view_type: vk::ImageViewType,
+        format: format::Format,
+        swizzle: format::Swizzle,
+        usage: image::Usage,
+        range: image::SubresourceRange,
+    ) -> Result<n::ImageView, image::ViewCreationError> {
```
I would have preferred adding a `Device::image_from_raw()` method, but in my case it is OpenXR that creates the Vulkan images and it hides some details needed to construct gfx images.

zarik5 commented:
The PR is ready. I tested it with this [example](https://github.com/zarik5/openxrs/blob/master/openxr/examples/vulkan-gfx.rs). I took the vulkan.rs example from [openxrs](https://github.com/Ralith/openxrs) and converted most of vulkan code to gfx-hal. I had some problems with crashing on close, fixed by using DropGuard on vk_instance and vk_device.

kvark reviewed (approved):
bors r+
